trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py

```diff
@@ -0,0 +1,1386 @@
+
+        Inputs:
+            self.engine.type                    string
+            aerodynamics.cd_min_clean           [Unitless]
+            cruise.altitude                     [m]
+                   delta_ISA                    [K]
+                   airspeed                     [m/s]
+                   thrust_fraction              [Unitless]
+            geometry.aspect_ratio               [Unitless]
+            wing_loading                        [N/m**2]
+
+        Outputs:
+            constraints.T_W                     [Unitless]
+                        W_S                     [N/m**2]
+
+        Properties Used:
+        """  
+
+        # Unpack inputs
+        if cruise_tag == 'max cruise':
+            altitude    = self.max_cruise.altitude 
+            delta_ISA   = self.max_cruise.delta_ISA
+            M           = self.max_cruise.mach 
+            throttle    = self.max_cruise.thrust_fraction 
+        else:
+            altitude    = self.cruise.altitude 
+            delta_ISA   = self.cruise.delta_ISA
+            M           = self.cruise.mach 
+            throttle    = self.cruise.thrust_fraction
+
+        eng_type  = self.engine.type
+        cd_min    = self.aerodynamics.cd_min_clean 
+        AR        = self.geometry.aspect_ratio
+        W_S       = self.wing_loading
+              
+        # Determine atmospheric properties at the altitude
+        planet      = SUAVE.Analyses.Planets.Planet()
+        atmosphere  = SUAVE.Analyses.Atmospheric.US_Standard_1976()
+        atmosphere.features.planet = planet.features
+
+        atmo_values = atmosphere.compute_values(altitude,delta_ISA)
+        rho         = atmo_values.density[0,0]
+        p           = atmo_values.pressure[0,0]
+        T           = atmo_values.temperature[0,0]
+        Vcr         = M * atmo_values.speed_of_sound[0,0]
+        q           = 0.5*rho*Vcr**2
+        g           = atmosphere.planet.sea_level_gravity
+   
+
+        T_W = np.zeros(len(W_S))
+        if eng_type != 'turbofan' and eng_type != 'turbojet':
+            P_W  = np.zeros(len(W_S))
+            etap = self.propeller.cruise_efficiency
+            if etap == 0:
+                raise ValueError('Warning: Set the propeller efficiency during cruise')
+        for i in range(len(W_S)):
+            CL = W_S[i]/q
+
+            # Calculate compressibility_drag
+            cd_comp   = self.compressibility_drag(M,CL) 
+            cd_comp_e = self.compressibility_drag(M,0)  
+            cd0       = cd_min + cd_comp  
+
+            # Calculate Oswald efficiency
+            if self.aerodynamics.oswald_factor == 0:
+                e = self.oswald_efficiency(cd_min+cd_comp_e,M)
+
+            k = 1/(np.pi*e*AR)
+            T_W[i] = q*cd0/W_S[i]+k*W_S[i]/q
+
+        # Convert thrust to power (for propeller-driven engines) and normalize the results wrt the Sea-level
+        if eng_type != 'turbofan' and eng_type != 'turbojet':
```

Comment:
Much of this code is copy and paste from other functions. Find ways to condense code.

This will help in several ways. It will reduce the size of this file, but most importantly it serves to reduce bugs.